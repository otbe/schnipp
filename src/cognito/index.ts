import { CognitoUserPoolTriggerEvent } from 'aws-lambda';
import { IHandler } from '../';

export type CognitoUserPoolHandler = IHandler<
  CognitoUserPoolTriggerEvent,
  CognitoUserPoolTriggerEvent
>;
