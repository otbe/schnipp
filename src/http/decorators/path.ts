import {
  RestControllerData,
  getControllerMetaData
} from './getControllerMetaData';

export const Path = (basePath: string = '/'): ClassDecorator => (
  target: any
) => {
  const metaData = getControllerMetaData(target) as RestControllerData;
  metaData.basePath = basePath;
};
