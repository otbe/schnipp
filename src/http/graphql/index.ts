export { GraphQLController } from './GraphQLController';
export { Query } from './decorators/query';
export { Mutation } from './decorators/mutation';
export { ResolveField } from './decorators/resolveField';
export { ResolveType } from './decorators/resolveType';
export { Header } from './decorators/header';
export { Meta } from './decorators/meta';
export { Use } from './decorators/use';
