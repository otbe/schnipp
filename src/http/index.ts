import {
  APIGatewayEvent,
  APIGatewayProxyResult,
  CustomAuthorizerEvent,
  CustomAuthorizerResult
} from 'aws-lambda';
import { IHandler } from '..';

export type APIGatewayHandler = IHandler<
  APIGatewayEvent,
  APIGatewayProxyResult
>;

export type AuthorizeHandler = IHandler<
  CustomAuthorizerEvent,
  CustomAuthorizerResult
>;

export { AuthPolicy, HttpVerb, Effect } from './utils/AuthPolicy';
export { createResponse } from './utils/createResponse';
export { MetaData } from './utils/MetaData';
export {
  DecoratedExceptionFilter,
  ExceptionFilter
} from './utils/ExceptionFilter';
export { Guard } from './utils/Guard';
export { DefaultExecutionContext } from './utils/ExecutionContext';
export { Catch } from './decorators/catch';
