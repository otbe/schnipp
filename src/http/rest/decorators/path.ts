import { inject } from 'simple-ts-di';
import {
  RestControllerData,
  getControllerMetaData
} from '../../decorators/getControllerMetaData';

export const Path = (basePath: string = '/'): ClassDecorator => (
  target: any
) => {
  inject()(target);
  const metaData = getControllerMetaData(target) as RestControllerData;
  metaData.basePath = basePath;
};
