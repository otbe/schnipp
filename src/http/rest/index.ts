export { RestController } from './RestController';

export {
  InjectedParam,
  MethodMetaData,
  ParamInjector
} from './factories/createHttpMethodDecorator';

export {
  GET,
  POST,
  DELETE,
  PUT,
  PATCH,
  OPTIONS,
  HEAD
} from './decorators/methods';

export { Path } from './decorators/path';
export { Param } from './decorators/param';
export { Query } from './decorators/query';
export { StatusCode } from './decorators/statusCode';
export { Context } from './decorators/context';
export { Body } from './decorators/body';
export { Header } from './decorators/header';
export { Meta } from './decorators/meta';
export { Use } from './decorators/use';

export { HttpException } from './exceptions/HttpException';
export { NotFoundException } from './exceptions/NotFoundException';
export { ForbiddenException } from './exceptions/ForbiddenException';
