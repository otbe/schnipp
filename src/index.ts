import { Context, Handler } from 'aws-lambda';
import { ContainedType, getFromContainer } from './container';

export type IHandler<E, R = void> = {
  handle(event: E, context: Context): Promise<R> | R;
};

let setupPromise: Promise<any> | null = null;
export const createHandler = <E, R>(
  handler: ContainedType<IHandler<E, R>>,
  setupFn?: () => Promise<any>
): Handler<E, R> => async (event: E, context: Context) => {
  context.callbackWaitsForEmptyEventLoop = false;

  if (setupFn) {
    if (setupPromise == null) {
      setupPromise = setupFn();
    }

    await setupPromise;
  }

  const handlerInstance = getFromContainer(handler);
  return await handlerInstance.handle(event, context);
};

export { useContainer } from './container';
