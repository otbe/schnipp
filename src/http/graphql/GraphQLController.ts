import { CONTAINER_INSTANCE_PROP, Container, Newable } from 'simple-ts-di';
import {
  APIGatewayProxyEvent,
  Context,
  APIGatewayProxyResult,
  APIGatewayEvent
} from 'aws-lambda';
import { graphqlLambda } from 'apollo-server-lambda';
import {
  APIGatewayHandler,
  Guard,
  MetaData,
  DecoratedExceptionFilter,
  ExceptionFilter
} from '..';
import {
  IResolverObject,
  makeExecutableSchema,
  SchemaDirectiveVisitor
} from 'graphql-tools';
import { readFile } from 'fs';
import { flatten } from 'lodash';
import { DefaultExecutionContext } from '../utils/ExecutionContext';
import {
  getControllerMetaData,
  GraphQLControllerData
} from '../decorators/getControllerMetaData';
import { GraphQLError } from 'graphql';
import { ResolverMethodMeta } from './decorators/method';

const { mergeTypes } = require('merge-graphql-schemas');

export abstract class GraphQLController implements APIGatewayHandler {
  private controllerData = getControllerMetaData(this
    .constructor as any) as GraphQLControllerData;

  private get container() {
    return (this as any)[CONTAINER_INSTANCE_PROP] as Container;
  }

  async handle(
    event: APIGatewayProxyEvent,
    lambdaContext: Context
  ): Promise<APIGatewayProxyResult> {
    const schema = await this._createSchema();
    const context = await this.createExecutionContext(event, lambdaContext);

    const handler = graphqlLambda({
      schema,
      context
    });

    return await new Promise<APIGatewayProxyResult>((resolve, reject) => {
      handler(event, lambdaContext, (err, output) => {
        if (err) {
          return reject(err);
        }

        const headers = this.controllerData.headers || {};
        output.headers = { ...output.headers, ...headers };

        resolve(output);
      });
    });
  }

  private async _createSchema() {
    const resolver = this.controllerData.resolver;
    if (resolver == null) {
      throw 'no resolver found';
    }

    const typesArray = await Promise.all(
      flatten(resolver.map(x => x.schemaPath).filter(Boolean) as Array<
        Array<string>
      >).map(
        x =>
          new Promise(resolve =>
            readFile(x, (err, res) => resolve(res.toString()))
          )
      )
    );

    const typeDefs = mergeTypes(typesArray, { all: true });

    const resolvers = resolver.reduce(
      (acc, curr) => {
        Object.keys(curr.resolverMap).forEach(typeName => {
          acc[typeName] = acc[typeName] || {};
          Object.keys(curr.resolverMap[typeName]).forEach(fieldName => {
            const resolverMeta = curr.resolverMap[typeName][fieldName];
            acc[typeName][fieldName] = this.createResolver(resolverMeta);
          });
        });

        return acc;
      },
      {} as { [typeName: string]: IResolverObject }
    );

    return makeExecutableSchema({
      typeDefs,
      resolvers,
      schemaDirectives: this.getSchemaDirectives()
    });
  }

  private createResolver(resolverMeta: ResolverMethodMeta<any>) {
    return async (source: any, args: any, context: any, info: any) => {
      const resolver = await this.container.get(resolverMeta.clazz!);

      const guards = [
        ...(resolverMeta.guards || []),
        ...this.controllerData.guards
      ];

      const metaData = {
        ...this.controllerData.metaData,
        ...(resolverMeta.metaData || {})
      };

      try {
        await this.processGuards(guards, context, metaData);

        return await resolver[resolverMeta.methodName!].call(
          resolver,
          source,
          args,
          context,
          info
        );
      } catch (exception) {
        const filters = [
          ...(resolverMeta.filters || []),
          ...this.controllerData.filters
        ];

        const mappedExceptionResponse = await this.processFilters(
          filters,
          exception,
          context,
          metaData
        );

        if (mappedExceptionResponse != null) {
          throw mappedExceptionResponse;
        }

        if (exception instanceof GraphQLError) {
          throw exception;
        }

        throw new GraphQLError('Internal Server Error');
      }
    };
  }

  private async processFilters(
    filters: Array<DecoratedExceptionFilter>,
    e: any,
    executionContext: DefaultExecutionContext | any,
    metaData: MetaData
  ) {
    const exceptionFilter = filters.find(x =>
      x.exceptions.some(y => e.name === y.name || e instanceof y)
    );

    if (exceptionFilter == null) {
      return;
    }

    const filter = await this.container.get<ExceptionFilter>(
      exceptionFilter.filter
    );

    return (await filter.catch(e, executionContext, metaData)) as
      | GraphQLError
      | undefined;
  }

  private async processGuards(
    guards: Array<Newable<Guard>>,
    executionContext: DefaultExecutionContext | any,
    metaData: MetaData
  ) {
    const canActivate = (await Promise.all(
      guards.map(guard =>
        this.container
          .get(guard)
          .then(guard => guard.canActivate(executionContext, metaData))
      )
    )).every(Boolean);

    if (!canActivate) {
      throw new GraphQLError('Forbidden');
    }
  }

  async createExecutionContext(
    event: APIGatewayEvent,
    lambdaContext: Context
  ): Promise<DefaultExecutionContext | any> {
    return { authorizer: event.requestContext.authorizer };
  }

  getSchemaDirectives(): {
    [name: string]: typeof SchemaDirectiveVisitor;
  } {
    return {};
  }
}
