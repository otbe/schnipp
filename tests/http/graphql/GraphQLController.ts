import 'reflect-metadata';
import {
  GraphQLController,
  Resolver,
  Query,
  ResolveField,
  Schema,
  Use,
  Header,
  Meta
} from '../../../src/http/graphql';
import { inject, Module, Bind, Container } from 'simple-ts-di';
import { join } from 'path';
import { HttpVerb, Guard, ExceptionFilter, Catch } from '../../../src/http';
import { HttpException } from '../../../src/http/rest';
import { GraphQLError } from 'graphql';

describe('GraphQLController.ts', () => {
  const context: any = {};
  const queryHandle = jest.fn();
  const fieldHandle = jest.fn();
  const fieldHandle2 = jest.fn();
  const guardHandle = jest.fn();
  const guardHandle2 = jest.fn();
  const filterHandle = jest.fn();
  const filterHandle2 = jest.fn();

  class Service {}

  class TestGuard implements Guard {
    canActivate(c, m) {
      return guardHandle(c, m);
    }
  }

  class TestGuard2 implements Guard {
    canActivate(c, m) {
      return guardHandle2(c, m);
    }
  }

  @Catch(HttpException)
  class TestFilter implements ExceptionFilter {
    catch(e, context, m) {
      return filterHandle(e, context, m);
    }
  }

  @Catch(HttpException)
  class TestFilter2 implements ExceptionFilter {
    catch(e, context, m) {
      return filterHandle2(e, context, m);
    }
  }

  @inject()
  @Header('Access-Control-Allow-Origin', '*')
  @Meta('foo', 'bar')
  @Use(TestGuard, TestFilter)
  class Test extends GraphQLController {
    async createExecutionContext() {
      return 5;
    }
  }

  @Resolver(Test)
  @Schema(join(__dirname, 'foos.gql'))
  class TestResolver {
    @Query('foos')
    foo(_, args, context) {
      return queryHandle(context);
    }

    @ResolveField('Foo', 'id')
    async fooId(foo) {
      await fieldHandle(foo);
      return foo.id + 1;
    }

    @ResolveField('Foo', 'name')
    @Meta('bar', 'baz')
    @Use(TestGuard2, TestFilter2)
    async fooName(foo) {
      await fieldHandle2(foo);
      return 'Hello ' + foo.name;
    }
  }

  class MyModule implements Module {
    init(bind: Bind) {
      bind(Service);
      bind(Test);
      bind(TestResolver);
      bind(TestGuard);
      bind(TestGuard2);
      bind(TestFilter);
      bind(TestFilter2);
    }
  }

  beforeEach(() => {
    queryHandle.mockClear();
    queryHandle.mockResolvedValue([{ id: 1, name: 'foo' }]);
    fieldHandle.mockClear();
    fieldHandle2.mockClear();
    filterHandle.mockReset();
    filterHandle2.mockReset();
    guardHandle.mockReset();
    guardHandle2.mockReset();
    guardHandle.mockResolvedValue(true);
    guardHandle2.mockResolvedValue(true);
  });

  it('should handle basic usage patterns ', async () => {
    const c = new Container(new MyModule());
    const h = await c.get(Test);
    const res = await h.handle(routeFor(query()), context);

    expect(guardHandle).toHaveBeenCalledWith(5, { foo: 'bar' });
    expect(queryHandle).toHaveBeenCalledWith(5);
    expect(fieldHandle).toHaveBeenCalled();
    expect(fieldHandle2).toHaveBeenCalled();
    expect(res).toMatchSnapshot();
  });

  it('should return forbidden error for entire API', async () => {
    guardHandle.mockResolvedValue(false);
    const c = new Container(new MyModule());
    const h = await c.get(Test);
    const res = await h.handle(routeFor(query()), context);

    expect(queryHandle).not.toHaveBeenCalled();
    expect(fieldHandle).not.toHaveBeenCalled();
    expect(fieldHandle2).not.toHaveBeenCalled();
    expect(res).toMatchSnapshot();
  });

  it('should return forbidden (null value) for specific field ', async () => {
    guardHandle2.mockResolvedValue(false);
    const c = new Container(new MyModule());
    const h = await c.get(Test);
    const res = await h.handle(routeFor(query()), context);

    expect(guardHandle2).toHaveBeenCalledWith(5, { bar: 'baz', foo: 'bar' });
    expect(queryHandle).toHaveBeenCalled();
    expect(fieldHandle).toHaveBeenCalled();
    expect(fieldHandle2).not.toHaveBeenCalled();
    expect(res).toMatchSnapshot();
  });

  it('should pass custom graphql errors to the client ', async () => {
    queryHandle.mockRejectedValueOnce(new GraphQLError('foo'));
    const c = new Container(new MyModule());
    const h = await c.get(Test);
    const res = await h.handle(routeFor(query()), context);

    expect(fieldHandle).not.toHaveBeenCalled();
    expect(fieldHandle2).not.toHaveBeenCalled();
    expect(res).toMatchSnapshot();
  });

  it('should map arbritary errors to graphql errors', async () => {
    queryHandle.mockRejectedValueOnce(new Error('foo'));
    const c = new Container(new MyModule());
    const h = await c.get(Test);
    const res = await h.handle(routeFor(query()), context);

    expect(fieldHandle).not.toHaveBeenCalled();
    expect(fieldHandle2).not.toHaveBeenCalled();
    expect(res).toMatchSnapshot();
  });

  it('should filter errors from controller on root types', async () => {
    filterHandle.mockResolvedValue(new GraphQLError('bar'));
    queryHandle.mockRejectedValue(new HttpException(400));
    const c = new Container(new MyModule());
    const h = await c.get(Test);
    const res = await h.handle(routeFor(query()), context);

    expect(filterHandle).toHaveBeenCalledWith(expect.any(HttpException), 5, {
      foo: 'bar'
    });
    expect(queryHandle).toHaveBeenCalled();
    expect(fieldHandle).not.toHaveBeenCalled();
    expect(fieldHandle2).not.toHaveBeenCalled();
    expect(res).toMatchSnapshot();
  });

  it('should filter errors from controller on abritrary types', async () => {
    filterHandle.mockResolvedValue(new GraphQLError('bar'));
    fieldHandle.mockRejectedValue(new HttpException(400));
    const c = new Container(new MyModule());
    const h = await c.get(Test);
    const res = await h.handle(routeFor(query()), context);

    expect(queryHandle).toHaveBeenCalled();
    expect(fieldHandle).toHaveBeenCalled();
    expect(fieldHandle2).toHaveBeenCalled();
    expect(res).toMatchSnapshot();
  });

  it('should filter errors from method', async () => {
    filterHandle2.mockResolvedValue(new GraphQLError('bar'));
    fieldHandle2.mockRejectedValue(new HttpException(400));
    const c = new Container(new MyModule());
    const h = await c.get(Test);
    const res = await h.handle(routeFor(query()), context);

    expect(filterHandle2).toHaveBeenCalledWith(expect.any(HttpException), 5, {
      bar: 'baz',
      foo: 'bar'
    });
    expect(queryHandle).toHaveBeenCalled();
    expect(fieldHandle).toHaveBeenCalled();
    expect(fieldHandle2).toHaveBeenCalled();
    expect(res).toMatchSnapshot();
  });
});

const query = () => JSON.stringify({ query: `{ foos { id name } }` });

function routeFor(body: string, requestContext = {}): any {
  return {
    body,
    httpMethod: HttpVerb.POST,
    requestContext
  };
}
