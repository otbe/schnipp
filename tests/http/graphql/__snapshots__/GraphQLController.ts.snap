// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQLController.ts should filter errors from controller on abritrary types 1`] = `
Object {
  "body": "{\\"data\\":{\\"foos\\":[null]},\\"errors\\":[{\\"message\\":\\"bar\\",\\"locations\\":[{\\"line\\":1,\\"column\\":10}],\\"path\\":[\\"foos\\",0,\\"id\\"]}]}",
  "headers": Object {
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 200,
}
`;

exports[`GraphQLController.ts should filter errors from controller on root types 1`] = `
Object {
  "body": "{\\"data\\":{\\"foos\\":null},\\"errors\\":[{\\"message\\":\\"bar\\",\\"locations\\":[{\\"line\\":1,\\"column\\":3}],\\"path\\":[\\"foos\\"]}]}",
  "headers": Object {
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 200,
}
`;

exports[`GraphQLController.ts should filter errors from method 1`] = `
Object {
  "body": "{\\"data\\":{\\"foos\\":[null]},\\"errors\\":[{\\"message\\":\\"Internal Server Error\\",\\"locations\\":[{\\"line\\":1,\\"column\\":10}],\\"path\\":[\\"foos\\",0,\\"id\\"]},{\\"message\\":\\"bar\\",\\"locations\\":[{\\"line\\":1,\\"column\\":13}],\\"path\\":[\\"foos\\",0,\\"name\\"]}]}",
  "headers": Object {
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 200,
}
`;

exports[`GraphQLController.ts should handle basic usage patterns  1`] = `
Object {
  "body": "{\\"data\\":{\\"foos\\":[{\\"id\\":\\"2\\",\\"name\\":\\"Hello foo\\"}]}}",
  "headers": Object {
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 200,
}
`;

exports[`GraphQLController.ts should map arbritary errors to graphql errors 1`] = `
Object {
  "body": "{\\"data\\":{\\"foos\\":null},\\"errors\\":[{\\"message\\":\\"Internal Server Error\\",\\"locations\\":[{\\"line\\":1,\\"column\\":3}],\\"path\\":[\\"foos\\"]}]}",
  "headers": Object {
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 200,
}
`;

exports[`GraphQLController.ts should pass custom graphql errors to the client  1`] = `
Object {
  "body": "{\\"data\\":{\\"foos\\":null},\\"errors\\":[{\\"message\\":\\"foo\\",\\"locations\\":[{\\"line\\":1,\\"column\\":3}],\\"path\\":[\\"foos\\"]}]}",
  "headers": Object {
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 200,
}
`;

exports[`GraphQLController.ts should return forbidden (null value) for specific field  1`] = `
Object {
  "body": "{\\"data\\":{\\"foos\\":[{\\"id\\":\\"2\\",\\"name\\":null}]},\\"errors\\":[{\\"message\\":\\"Forbidden\\",\\"locations\\":[{\\"line\\":1,\\"column\\":13}],\\"path\\":[\\"foos\\",0,\\"name\\"]}]}",
  "headers": Object {
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 200,
}
`;

exports[`GraphQLController.ts should return forbidden error for entire API 1`] = `
Object {
  "body": "{\\"data\\":{\\"foos\\":null},\\"errors\\":[{\\"message\\":\\"Forbidden\\",\\"locations\\":[{\\"line\\":1,\\"column\\":3}],\\"path\\":[\\"foos\\"]}]}",
  "headers": Object {
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 200,
}
`;
